### PR TITLE
Add support for Hetzner Private Networking

### LINKED ISSUE #661: Support private hetzner networking

### PR DESCRIPTION
**Pull request by kron4eg**
Fixes #661

```release-note
Add support for Hetzner Private Networking
```

### CODE REVIEW

**Comment by kron4eg** [comment]
WIP because we also need to add hetzner private networking support to machine-controller

**Comment by xmudrii** [comment]
This PR is currently blocked as Hetzner CCM doesn't support private networks

https://github.com/hetznercloud/hcloud-cloud-controller-manager/issues/19
https://github.com/hetznercloud/hcloud-cloud-controller-manager/pull/20

There is a WIP PR to implement Hetzner Networks for machine-controller

https://github.com/kubermatic/machine-controller/pull/627

**Comment by xmudrii** [comment]
This issue is not anymore blocked as we have a new release of both machine-controller and Hetzner CCM.

/remove-lifecycle blocked

File: pkg/apis/kubeone/types.go [hunk 1]
```diff
@@ -119,6 +119,10 @@ type ClusterNetworkConfig struct {
 	ServiceDomainName string `json:"serviceDomainName"`
 	NodePortRange     string `json:"nodePortRange"`
 	CNI               *CNI   `json:"cni,omitempty"`
+
+	// TODO(xmudrii): Revisit do we want to put this here.
+	// +optional
+	NetworkID string `json:"networkID,omitempty"`
```
Comment (xmudrii) [inline pkg/apis/kubeone/types.go]: Do we want this field to be here? If not, where should I put it?

File: pkg/apis/kubeone/types.go [hunk 2]
```diff
@@ -119,6 +119,10 @@ type ClusterNetworkConfig struct {
 	ServiceDomainName string `json:"serviceDomainName"`
 	NodePortRange     string `json:"nodePortRange"`
 	CNI               *CNI   `json:"cni,omitempty"`
+
+	// TODO(xmudrii): Revisit do we want to put this here.
+	// +optional
```
Comment (xmudrii) [inline pkg/apis/kubeone/types.go]: Should I keep the `+optional` tag? If yes, should I make this a pointer to string?

File: pkg/templates/externalccm/hetzner.go [hunk 3]
```diff
@@ -183,6 +193,11 @@ func hetznerDeployment() *appsv1.Deployment {
 										},
 									},
 								},
+								{
+									Name: "HCLOUD_NETWORK",
+									// TODO(xmudrii): Should we keep the network ID/name in the secret?
+									Value: networkID,
```
Comment (xmudrii) [inline pkg/templates/externalccm/hetzner.go]: Should this be in a secret or this approach is okay?

**Comment by xmudrii** [comment]
/hold
I haven't pushed and tested a fix for CCM flags

File: examples/terraform/hetzner/main.tf [hunk 4]
```diff
@@ -19,6 +19,24 @@ resource "hcloud_ssh_key" "kubeone" {
...
+resource "hcloud_network" "net" {
+  name     = var.cluster_name
+  ip_range = "192.168.0.0/16"
+}
+
+resource "hcloud_network_subnet" "kubeone" {
+  network_id   = hcloud_network.net.id
+  type         = "server"
+  network_zone = "eu-central"
+  ip_range     = "192.168.0.0/24"
```
Comment (kron4eg) [inline examples/terraform/hetzner/main.tf]: and this

File: pkg/templates/externalccm/hetzner.go (hunk 3 above)
Comment (kron4eg) [reply pkg/templates/externalccm/hetzner.go]: I don't know why networkID should be treated as a secret. So it's OK like this.

File: pkg/apis/kubeone/types.go (hunk 2 above)
Comment (kron4eg) [reply pkg/apis/kubeone/types.go]: We should follow here what machine-controller uses (which is []string)

**Pull request merged**